### PR TITLE
Add Rng interface and an example that uses it.

### PR DESCRIPTION
**Pull request by tstellanova**
Enables use of the RNG (random number generator) peripheral available on some stm32f4 chips. 

This PR includes an example openocd configuration for running the example.

### CODE REVIEW

**Comment by tstellanova** [comment]
I'll look into the CI failures. Looks like it's failing on some processors that don't have RNG

File: src/rng.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    SeedError,
+}
+
+
+pub trait RngExt {
+    fn constrain(self) -> Rng;
+}
+
+impl RngExt for RNG {
+
```
Comment (therealprof) [inline src/rng.rs]:
```suggestion
```

**Comment by therealprof** [comment]
Thank you for this PR. It looks very good in general. I'm not too thrilled about the OpenOCD stuff, this really shouldn't be in here. I'll add a few more comments and nits inline.

File: examples/rng-display.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+//! Note that this example requires the `--release` build flag because it is too
+//! large to fit in the default `memory.x` file provided with this crate.
+
+#![no_std]
+#![no_main]
+
+extern crate cortex_m;
+extern crate cortex_m_rt as rt;
+extern crate panic_semihosting;
+extern crate stm32f4xx_hal as hal;
```
Comment (therealprof) [inline examples/rng-display.rs]: Can we please make this edition 2018?

File: src/rng.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    /// Enable RNG_CLK and the RNG peripheral.
+    /// Note that clocks must already be configured
+    /// such that RNG_CLK is not less than 1/16 HCLK,
+    /// otherwise all reads of the RNG will return
+    /// a ClockError (CECS error)
+    fn constrain(self) -> Rng {
```
Comment (therealprof) [inline src/rng.rs]: We should probably take the clock configuration here and make sure the prerequisites for the clock speeds are met.

File: src/rng.rs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                    Ok(val as $type)
+                }
+
+                /// Fills buffer with random values, or return error
+                fn fill(&mut self, buffer: &mut [$type]) -> Result<(), ErrorKind> {
+                    const BATCH_SIZE: usize = 4 / mem::size_of::<$type>();
+                    let mut i = 0_usize;
+                    while i < buffer.len() {
+                        let random_word = self.next()?;
+                        let bytes: [$type; BATCH_SIZE] = unsafe { mem::transmute(random_word) };
```
Comment (therealprof) [inline src/rng.rs]: Not too hot about the transmute here. Probably should find a safer way to do it.

File: src/rng.rs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+impl rng::Read for Rng {
+    type Error = ErrorKind;
+
+    fn read(&mut self, buffer: &mut [u8]) -> Result<(), Self::Error> {
+        self.fill(buffer)
+    }
+}
+
+macro_rules! rng_core {
```
Comment (therealprof) [inline src/rng.rs]: Can't we simply use an associated type here instead of repeating the impl with a macro?

File: src/rng.rs (hunk 4 above)
Comment (TeXitoi) [reply src/rng.rs]: https://doc.rust-lang.org/beta/std/primitive.u64.html#method.to_ne_bytes

So:
```suggestion
                        let bytes = random_word.to_ne_bytes();
```

File: src/rng.rs (hunk 5 above)
Comment (TeXitoi) [reply src/rng.rs]: I'm not sure I understand your proposition. Do you want this macro be used:
```rust
rng_core!(usize);
rng_core(u32);
rng_core(u16)
rng_code(u8); 
```

File: src/rng.rs (hunk 4 above)
Comment (TeXitoi) [reply src/rng.rs]: Maybe a bit more complex, but `from_ne_bytes()` and `to_ne_bytes()` looks like good tools.

File: src/rng.rs (hunk 5 above)
Comment (therealprof) [reply src/rng.rs]: No, I mean move as much as possible out of such a macro, e.g. by turning the substituted type into an associated type.

**Comment by tstellanova** [comment]
Taking another direction here and implementing the no_std interface of `rand_core`...stay tuned.

**Comment by therealprof** [comment]
@tstellanova Cool, looking forward to it.

**Comment by therealprof** [comment]
@tstellanova Looks excellent. I'm going to try it right away. Meanwhile would you please mention this new feature in https://github.com/stm32-rs/stm32f4xx-hal/blob/master/CHANGELOG.md and make sure to run `rustfmt` on it?

File: examples/rng-display.rs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+#[cfg(debug_assertions)]
+use panic_semihosting as _;
+
+use cortex_m_rt::ExceptionFrame;
+use cortex_m_rt::{entry, exception};
+use embedded_graphics::{fonts::Font6x8, prelude::*};
+
+use ssd1306::{prelude::*, Builder as SSD1306Builder};
+
+use hal::rng::RngExt;
```
Comment (therealprof) [inline examples/rng-display.rs]: Please move this to the prelude so everyone who already uses the prelude can just use it.

**Comment by therealprof** [comment]
I've tested it and it works a treat for me.

@TeXitoi Any thoughts from your side?

**Comment by tstellanova** [comment]
The prelude change failed CI, I think because not all chips have the RNG peripheral.  I can either remove from prelude (similar to usb_fs) or add cfg features  in the prelude.  Any preferences?

**Comment by therealprof** [comment]
Can you squash the commits a little to get rid of the failing intermediate changes? Then I would say this is ready to go in.

**Pull request merged**